File: library/src/main/java/com/dmanchester/playfop/japi/PlayFop.java
package com.dmanchester.playfop.japi;

import java.io.OutputStream;

import org.apache.fop.apps.Fop;

import play.twirl.api.Xml;

/**
 * The primary entry point into PlayFOP for Java applications.
 */
public interface PlayFop {

    /**
     * Processes XSL-FO with Apache FOP. Generates output in the specified
     * format.
     *
     * @param xslfo the XSL-FO to process
     * @param outputFormat the format to generate
     * @return the Apache FOP output
     */
    public byte[] process(Xml xslfo, String outputFormat);

    /**
     * Processes XSL-FO with Apache FOP, applying the processing options.
     * Generates output in the specified format.
     *
     * @param xslfo the XSL-FO to process
     * @param outputFormat the format to generate
     * @param processOptions the processing options
     * @return the Apache FOP output
     */
    public byte[] process(Xml xslfo, String outputFormat, ProcessOptions processOptions);

    /**
     * Creates a new <code>Fop</code> instance. Sets it up to save output to the
     * supplied <code>OutputStream</code> in the supplied format.
     * <p>
     * <b>Note:</b> This method is offered primarily for client code to
     * interrogate the Apache FOP environment (for example, to determine
     * available fonts). Code wishing to process XSL-FO with Apache FOP should
     * rely on the <code>process</code> method instead.
     *
     * @param outputFormat the format the <code>Fop</code> should generate
     * @param output the <code>OutputStream</code> to which the <code>Fop</code>
     *        should save output
     * @return the <code>Fop</code>
     */
    public Fop newFop(String outputFormat, OutputStream output);

    /**
     * Creates a new <code>Fop</code> instance, configuring it with the
     * processing options. Sets it up to save output to the supplied
     * <code>OutputStream</code> in the supplied format.
     * <p>
     * <b>Note:</b> This method is offered primarily for client code to
     * interrogate the Apache FOP environment (for example, to determine
     * available fonts). Code wishing to process XSL-FO with Apache FOP should
     * rely on a <code>process</code> method instead.
     *
     * @param outputFormat the format the <code>Fop</code> should generate
     * @param output the <code>OutputStream</code> to which the <code>Fop</code>
     *        should save output
     * @param processOptions the processing options
     * @return the <code>Fop</code>
     */
    public Fop newFop(String outputFormat, OutputStream output, ProcessOptions processOptions);
}

File: library/src/main/scala/com/dmanchester/playfop/api_s/PlayFop.scala
package com.dmanchester.playfop.api_s

import java.io.OutputStream

import org.apache.fop.apps.FOUserAgent
import org.apache.fop.apps.Fop

import play.twirl.api.Xml

/** The primary entry point into PlayFOP for Scala applications.
  */
trait PlayFop {

  /** Processes XSL-FO with Apache FOP, optionally auto-detecting fonts (for PDF
    * output) and/or applying a code block to the `FOUserAgent`. Generates
    * output in the specified format.
    *
    * @tparam U the return type of `foUserAgentBlock` (typically inferred, as
    *           opposed to explicitly specified)
    * @param xslfo the XSL-FO to process
    * @param outputFormat the format to generate
    * @param autoDetectFontsForPDF whether to auto-detect fonts
    * @param foUserAgentBlock the code block for the `FOUserAgent`
    * @return the Apache FOP output
    */
  def process[U](xslfo: Xml, outputFormat: String,
      autoDetectFontsForPDF: Boolean = false,
      foUserAgentBlock: (FOUserAgent => U) = {_: FOUserAgent => }): Array[Byte]

  /** Creates a new `Fop` instance, optionally auto-detecting fonts (for PDF
    * output) and/or applying a code block to the `FOUserAgent`. Sets up the
    * `Fop` to save output to the supplied `OutputStream` in the supplied format.
    *
    * '''Note:''' This method is offered primarily for client code to
    * interrogate the Apache FOP environment (for example, to determine
    * available fonts). Code wishing to process XSL-FO with Apache FOP should
    * rely on the `process` method instead.
    *
    * @tparam U the return type of `foUserAgentBlock` (typically inferred, as
    *           opposed to explicitly specified)
    * @param outputFormat the format the `Fop` should generate
    * @param output the `OutputStream` to which the `Fop` should save output
    * @param autoDetectFontsForPDF whether to auto-detect fonts
    * @param foUserAgentBlock the code block for the `Fop`'s `FOUserAgent`
    * @return the `Fop`
    */
  def newFop[U](outputFormat: String, output: OutputStream,
      autoDetectFontsForPDF: Boolean = false,
      foUserAgentBlock: (FOUserAgent => U) = {_: FOUserAgent => }): Fop
}